digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/csvDataImporter/TrafficSignCsvImporter.scala

```diff
@@ -89,20 +235,155 @@ class TrafficSignCsvImporter(roadLinkServiceImpl: RoadLinkService, eventBusImpl:
+    val optEndDate = getPropertyValueOption(parsedRow, "endDate").asInstanceOf[Option[String]]
+    val isValidDate = optTrafficSignType match {
+      case Some(signType) if (!isBlank(optLifeCycle.toString) && temporaryDevices.contains(optLifeCycle)) || TrafficSignType.applyNewLawCode(signType) == RoadWorks =>
+        (optStartDate, optEndDate) match {
+          case (Some(startDate), Some(endDate)) =>
+            try {
+              val startDateFormat = dateFormatter.parseDateTime(startDate)
+              val endDateFormat = dateFormatter.parseDateTime(endDate)
+              endDateFormat.isAfter(startDateFormat) || endDateFormat.isEqual(startDateFormat)
+            } catch {
+              case _: Throwable => false
+            }
+          case (_, _) => false
+        }
+      case _ => true
+    }
+
+    if (!isValidDate) {
+      errorMessages = (errorMessages._1 ++ List("Invalid dates"), Seq())
+    }
+    /* end date validations */
+
+    /* start direction validations */
+    val optTrafficDirection = getPropertyValueOption(parsedRow, "bearing").asInstanceOf[Option[String]]
+    val directionValidator = optTrafficDirection match {
+      case Some(direction) if direction.trim.nonEmpty => true
+      case Some(_) if hasRoadName => true
+      case None if hasRoadName => true
+      case _ =>
+        errorMessages = (errorMessages._1 ++ List("Invalid traffic sign direction"), Seq())
+        false
+    }
+    /* end direction validations */
+
+    /* start additional panels type validations */
+    val minAdditionalPanels = 1
+    val maxAdditionalPanels = 3
+    val additionalPanelsValidator = optTrafficSignType match {
+      case Some(sType) if sType.trim.nonEmpty =>
+        val panelVerification = (minAdditionalPanels to maxAdditionalPanels).map { index =>
+          getPropertyValueOption(parsedRow, "additionalPanelType" + index).asInstanceOf[Option[String]] match {
+            case Some(panelType) if panelType.trim.nonEmpty && TrafficSignType.applyNewLawCode(panelType).group.value != AdditionalPanels.value =>
+              errorMessages = (errorMessages._1 ++ List("Invalid additional panel type on lisakilpi " + index), Seq())
+              false
+            case _ => true
+          }
+        }
+        panelVerification.count(_ == false) < 1
+      case _ => true
+    }
+    /* end additional panels type validations */
+
+    /* start lane type validations */
+    val optLaneType = getPropertyValueOption(parsedRow, "laneType").asInstanceOf[Option[Int]]
+    val optLane = getPropertyValueOption(parsedRow, "lane").asInstanceOf[Option[String]]
+    val lanesValidator = (optLaneType, optLane) match {
+      case (_, Some(lane)) if lane.trim.nonEmpty && !lane.matches("^([1-3][1-9])$") =>
+        errorMessages = (errorMessages._1 ++ List("Invalid lane"), Seq())
+        false
+      case (Some(laneType), Some(lane)) if laneType != 999 && lane.trim.nonEmpty =>
```

Comment:
If I'm checking that correctly the Type 999 never exists, the value for the Unknown LaneType is 99.
And we can switch this code to something like:
```Scala
case (Some(laneType), Some(lane)) if laneType != LaneType.Unknown.value && lane.trim.nonEmpty =>

Reply:
Done